File: smtp/typescript/examples/connection-pooling.ts
/**
 * Connection Pooling Example
 *
 * This example demonstrates how to use connection pooling to efficiently
 * send batch emails, monitor pool status, and optimize for high throughput.
 */

import {
  smtpClient,
  EmailBuilder,
  TlsMode,
  PoolStatus,
  SmtpMetrics,
} from '@integrations/smtp';

/**
 * Example 1: Basic connection pooling for batch email sending
 */
async function basicPoolingExample() {
  console.log('=== Basic Connection Pooling Example ===\n');

  // Create client with connection pooling enabled
  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    // Configure connection pool
    .poolConfig({
      maxConnections: 10,        // Maximum concurrent connections
      minIdleConnections: 2,     // Keep at least 2 connections ready
      maxIdleTime: 60000,        // Close idle connections after 60 seconds
      acquireTimeout: 10000,     // Wait up to 10 seconds for a connection
      validationInterval: 30000, // Validate connections every 30 seconds
    })
    .build();

  try {
    // Generate 50 test emails
    const emails = Array.from({ length: 50 }, (_, i) =>
      new EmailBuilder()
        .from('sender@example.com')
        .to(`user${i + 1}@example.com`)
        .subject(`Email #${i + 1}`)
        .text(`This is test email number ${i + 1}`)
        .build()
    );

    console.log(`Sending ${emails.length} emails using connection pool...`);
    const startTime = Date.now();

    // Send batch - the pool will automatically manage connections
    const result = await client.sendBatch(emails);

    const duration = Date.now() - startTime;

    console.log('\nResults:');
    console.log(`  Total: ${emails.length}`);
    console.log(`  Successful: ${result.successful.length}`);
    console.log(`  Failed: ${result.failed.length}`);
    console.log(`  Duration: ${duration}ms`);
    console.log(`  Throughput: ${(emails.length / (duration / 1000)).toFixed(2)} emails/sec`);

    // Report any failures
    if (result.failed.length > 0) {
      console.log('\nFailed emails:');
      result.failed.forEach(({ email, error }) => {
        console.log(`  - ${email.to[0].address}: ${error.message}`);
      });
    }

  } finally {
    await client.close();
  }
}

/**
 * Example 2: Monitoring pool status
 */
async function poolMonitoringExample() {
  console.log('\n=== Pool Monitoring Example ===\n');

  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    .poolConfig({
      maxConnections: 5,
      minIdleConnections: 1,
      maxIdleTime: 30000,
    })
    .build();

  try {
    // Monitor pool status periodically
    const monitorInterval = setInterval(async () => {
      const status: PoolStatus = await client.getPoolStatus();

      console.log('Pool Status:');
      console.log(`  Active: ${status.activeConnections}`);
      console.log(`  Idle: ${status.idleConnections}`);
      console.log(`  Total: ${status.totalConnections}`);
      console.log(`  Pending: ${status.pendingAcquisitions}`);
      console.log(`  Utilization: ${(status.utilization * 100).toFixed(1)}%`);
      console.log('');
    }, 2000);

    // Send emails in waves to see pool dynamics
    for (let wave = 1; wave <= 3; wave++) {
      console.log(`\n--- Wave ${wave} ---`);

      const emails = Array.from({ length: 15 }, (_, i) =>
        new EmailBuilder()
          .from('sender@example.com')
          .to(`user${wave * 15 + i}@example.com`)
          .subject(`Wave ${wave} Email ${i + 1}`)
          .text('Test email')
          .build()
      );

      await client.sendBatch(emails);

      // Wait between waves
      if (wave < 3) {
        await new Promise(resolve => setTimeout(resolve, 5000));
      }
    }

    // Stop monitoring
    clearInterval(monitorInterval);

  } finally {
    await client.close();
  }
}

/**
 * Example 3: High-performance batch sending with metrics
 */
async function highPerformanceExample() {
  console.log('\n=== High-Performance Batch Sending ===\n');

  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    // Optimize pool for high throughput
    .poolConfig({
      maxConnections: 20,        // More connections for parallel sends
      minIdleConnections: 5,     // Keep connections warm
      maxIdleTime: 120000,       // Keep connections alive longer
      acquireTimeout: 5000,      // Fail fast if pool is saturated
      validationInterval: 60000,
    })
    // Add rate limiting to avoid overwhelming the server
    .rateLimitConfig({
      maxPerSecond: 50,
      maxPerMinute: 2000,
      maxPerHour: 50000,
      onLimitBehavior: 'throttle' as const,
    })
    // Enable retries for transient failures
    .retryConfig({
      maxAttempts: 3,
      initialDelay: 1000,
      maxDelay: 10000,
      backoffMultiplier: 2.0,
      retryableErrors: [
        'connection_error',
        'timeout',
        'temporary_error',
      ],
    })
    // Enable circuit breaker to stop trying if server is down
    .circuitBreakerConfig({
      failureThreshold: 10,
      successThreshold: 5,
      timeout: 60000,
      halfOpenMaxAttempts: 3,
    })
    .build();

  try {
    // Generate a large batch of emails
    const batchSize = 1000;
    const emails = Array.from({ length: batchSize }, (_, i) =>
      new EmailBuilder()
        .from('bulk@example.com')
        .to(`customer${i + 1}@example.com`)
        .subject('Monthly Newsletter')
        .html(`
          <h1>Newsletter ${i + 1}</h1>
          <p>Thank you for being our customer!</p>
        `)
        .build()
    );

    console.log(`Sending ${batchSize} emails with optimized settings...`);

    // Monitor progress
    let completed = 0;
    const startTime = Date.now();

    // Send in chunks for better monitoring
    const chunkSize = 100;
    const chunks = [];
    for (let i = 0; i < emails.length; i += chunkSize) {
      chunks.push(emails.slice(i, i + chunkSize));
    }

    console.log(`Split into ${chunks.length} chunks of ${chunkSize} emails each\n`);

    const results = [];
    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i];
      const chunkStart = Date.now();

      const result = await client.sendBatch(chunk);
      results.push(result);

      completed += chunk.length;
      const chunkDuration = Date.now() - chunkStart;
      const overallDuration = Date.now() - startTime;
      const progress = (completed / batchSize) * 100;

      console.log(`Chunk ${i + 1}/${chunks.length}:`);
      console.log(`  Sent: ${result.successful.length}/${chunk.length}`);
      console.log(`  Duration: ${chunkDuration}ms`);
      console.log(`  Progress: ${progress.toFixed(1)}%`);
      console.log(`  Overall rate: ${(completed / (overallDuration / 1000)).toFixed(2)} emails/sec`);

      // Get pool status
      const poolStatus = await client.getPoolStatus();
      console.log(`  Pool: ${poolStatus.activeConnections} active, ${poolStatus.idleConnections} idle`);
      console.log('');
    }

    // Final statistics
    const totalDuration = Date.now() - startTime;
    const totalSuccessful = results.reduce((sum, r) => sum + r.successful.length, 0);
    const totalFailed = results.reduce((sum, r) => sum + r.failed.length, 0);

    console.log('\n=== Final Results ===');
    console.log(`Total emails: ${batchSize}`);
    console.log(`Successful: ${totalSuccessful}`);
    console.log(`Failed: ${totalFailed}`);
    console.log(`Total duration: ${(totalDuration / 1000).toFixed(2)}s`);
    console.log(`Average throughput: ${(totalSuccessful / (totalDuration / 1000)).toFixed(2)} emails/sec`);

    // Get final metrics
    const metrics: SmtpMetrics = await client.getMetrics();
    console.log('\n=== Connection Metrics ===');
    console.log(`Total connections created: ${metrics.connectionsCreated}`);
    console.log(`Connections reused: ${metrics.connectionsReused}`);
    console.log(`Connection errors: ${metrics.connectionErrors}`);
    console.log(`Average connection time: ${metrics.averageConnectionTime.toFixed(0)}ms`);

    console.log('\n=== Send Metrics ===');
    console.log(`Total sends: ${metrics.totalSends}`);
    console.log(`Successful sends: ${metrics.successfulSends}`);
    console.log(`Failed sends: ${metrics.failedSends}`);
    console.log(`Average send time: ${metrics.averageSendTime.toFixed(0)}ms`);
    console.log(`Retries: ${metrics.retriesAttempted}`);

  } finally {
    await client.close();
  }
}

/**
 * Example 4: Connection pool with graceful shutdown
 */
async function gracefulShutdownExample() {
  console.log('\n=== Graceful Shutdown Example ===\n');

  const client = smtpClient()
    .host('smtp.example.com')
    .port(587)
    .credentials('user@example.com', 'your-password')
    .tlsMode(TlsMode.StartTls)
    .poolConfig({
      maxConnections: 5,
      minIdleConnections: 1,
    })
    .build();

  // Handle shutdown signals
  let isShuttingDown = false;

  const shutdown = async () => {
    if (isShuttingDown) return;
    isShuttingDown = true;

    console.log('\nReceived shutdown signal...');
    console.log('Waiting for in-flight emails to complete...');

    // The close() method will wait for active connections to finish
    await client.close();

    console.log('All connections closed gracefully.');
    process.exit(0);
  };

  process.on('SIGINT', shutdown);
  process.on('SIGTERM', shutdown);

  try {
    // Simulate long-running email sending process
    console.log('Starting continuous email sending...');
    console.log('Press Ctrl+C to trigger graceful shutdown\n');

    let emailCount = 0;
    while (!isShuttingDown) {
      const email = new EmailBuilder()
        .from('sender@example.com')
        .to(`user${emailCount}@example.com`)
        .subject('Continuous Send Test')
        .text('Test email')
        .build();

      await client.send(email);
      emailCount++;

      console.log(`Sent email #${emailCount}`);

      // Wait 2 seconds between sends
      await new Promise(resolve => setTimeout(resolve, 2000));
    }

  } catch (error) {
    if (!isShuttingDown) {
      console.error('Error:', error);
    }
  }
}

/**
 * Run all examples
 */
async function main() {
  try {
    await basicPoolingExample();
    await poolMonitoringExample();
    await highPerformanceExample();

    // Uncomment to test graceful shutdown (runs indefinitely)
    // await gracefulShutdownExample();

  } catch (error) {
    console.error('Example failed:', error);
    process.exit(1);
  }
}

// Run examples
if (require.main === module) {
  main();
}

export {
  basicPoolingExample,
  poolMonitoringExample,
  highPerformanceExample,
  gracefulShutdownExample,
};

File: smtp/rust/examples/connection_pooling.rs
//! Connection Pooling Example
//!
//! This example demonstrates how to:
//! - Configure connection pool settings
//! - Send batch emails efficiently using pooled connections
//! - Monitor pool status and statistics
//! - Handle connection reuse and lifecycle

use smtp_integration::{
    SmtpClient, SmtpConfig, Email, EmailAddress, PoolConfig, SmtpError,
};
use std::time::Duration;
use tokio::time::sleep;

#[tokio::main]
async fn main() -> Result<(), SmtpError> {
    // Configure connection pooling for efficient batch sending
    let pool_config = PoolConfig::builder()
        .max_connections(10)           // Maximum number of pooled connections
        .min_idle_connections(2)        // Minimum number of idle connections to maintain
        .max_idle_time(Duration::from_secs(300))  // Close idle connections after 5 minutes
        .connection_timeout(Duration::from_secs(30))  // Timeout for establishing connections
        .idle_timeout(Duration::from_secs(60))    // Timeout for idle connections
        .test_on_checkout(true)         // Verify connection health before use
        .build()?;

    // Configure the SMTP client with pooling
    let config = SmtpConfig::builder()
        .host("smtp.example.com")
        .port(587)
        .username("user@example.com")
        .password("your-password")
        .use_tls(true)
        .use_starttls(true)
        .pool_config(pool_config)
        .build()?;

    println!("Creating SMTP client with connection pooling...");
    let client = SmtpClient::new(config)?;

    // Display initial pool status
    print_pool_status(&client);

    // Example 1: Send a batch of emails
    println!("\n=== Sending batch emails ===");
    send_batch_emails(&client).await?;

    // Display pool status after batch
    print_pool_status(&client);

    // Example 2: Send emails concurrently
    println!("\n=== Sending emails concurrently ===");
    send_concurrent_emails(&client).await?;

    // Display pool status after concurrent sends
    print_pool_status(&client);

    // Example 3: Monitor pool during sustained load
    println!("\n=== Sustained load test ===");
    sustained_load_test(&client).await?;

    // Final pool status
    print_pool_status(&client);

    // Graceful shutdown
    println!("\n=== Shutting down ===");
    client.close_pool().await?;
    println!("Connection pool closed gracefully");

    Ok(())
}

/// Send a batch of emails sequentially
async fn send_batch_emails(client: &SmtpClient) -> Result<(), SmtpError> {
    let recipients = vec![
        ("alice@example.com", "Alice"),
        ("bob@example.com", "Bob"),
        ("charlie@example.com", "Charlie"),
        ("diana@example.com", "Diana"),
        ("eve@example.com", "Eve"),
    ];

    for (i, (email, name)) in recipients.iter().enumerate() {
        let email_msg = Email::builder()
            .from(EmailAddress::new("newsletter@example.com", Some("Newsletter Team")))
            .to(EmailAddress::new(email, Some(name)))
            .subject(format!("Weekly Newsletter #{}", i + 1))
            .text_body(format!(
                "Hello {},\n\nWelcome to this week's newsletter!\n\nBest regards,\nThe Team",
                name
            ))
            .build()?;

        print!("  Sending to {}... ", name);
        match client.send(&email_msg).await {
            Ok(response) => println!("OK (Message ID: {:?})", response.message_id),
            Err(e) => println!("FAILED: {}", e),
        }

        // Small delay between sends
        sleep(Duration::from_millis(100)).await;
    }

    println!("Batch complete: {} emails sent", recipients.len());
    Ok(())
}

/// Send emails concurrently using the connection pool
async fn send_concurrent_emails(client: &SmtpClient) -> Result<(), SmtpError> {
    let recipients = vec![
        ("user1@example.com", "User 1"),
        ("user2@example.com", "User 2"),
        ("user3@example.com", "User 3"),
        ("user4@example.com", "User 4"),
        ("user5@example.com", "User 5"),
        ("user6@example.com", "User 6"),
        ("user7@example.com", "User 7"),
        ("user8@example.com", "User 8"),
    ];

    println!("  Sending {} emails concurrently...", recipients.len());

    // Create futures for all email sends
    let send_futures: Vec<_> = recipients
        .into_iter()
        .map(|(email, name)| {
            let client = client.clone();
            async move {
                let email_msg = Email::builder()
                    .from(EmailAddress::new("notifications@example.com", Some("Notification Service")))
                    .to(EmailAddress::new(email, Some(name)))
                    .subject("Important Update")
                    .text_body(format!("Hello {},\n\nThis is an important update.\n", name))
                    .build()?;

                client.send(&email_msg).await
            }
        })
        .collect();

    // Execute all sends concurrently
    let results = futures::future::join_all(send_futures).await;

    // Count successes and failures
    let successful = results.iter().filter(|r| r.is_ok()).count();
    let failed = results.iter().filter(|r| r.is_err()).count();

    println!("  Concurrent sends complete:");
    println!("    Successful: {}", successful);
    println!("    Failed: {}", failed);

    Ok(())
}

/// Simulate sustained load to observe pool behavior
async fn sustained_load_test(client: &SmtpClient) -> Result<(), SmtpError> {
    println!("  Running sustained load test (30 emails over 10 seconds)...");

    let start = std::time::Instant::now();
    let mut sent_count = 0;

    // Send emails for 10 seconds
    while start.elapsed() < Duration::from_secs(10) {
        let email = Email::builder()
            .from(EmailAddress::new("loadtest@example.com", Some("Load Test")))
            .to(EmailAddress::new("recipient@example.com", Some("Recipient")))
            .subject(format!("Load Test Email #{}", sent_count + 1))
            .text_body("This is a load test email.")
            .build()?;

        if let Ok(_) = client.send(&email).await {
            sent_count += 1;
            if sent_count % 10 == 0 {
                print_pool_status(client);
            }
        }

        sleep(Duration::from_millis(333)).await; // ~3 emails per second
    }

    let duration = start.elapsed();
    let rate = sent_count as f64 / duration.as_secs_f64();

    println!("  Sustained load test complete:");
    println!("    Emails sent: {}", sent_count);
    println!("    Duration: {:.2}s", duration.as_secs_f64());
    println!("    Rate: {:.2} emails/second", rate);

    Ok(())
}

/// Print current pool status
fn print_pool_status(client: &SmtpClient) {
    if let Some(status) = client.pool_status() {
        println!("\n  Pool Status:");
        println!("    Total connections: {}", status.total_connections);
        println!("    Active connections: {}", status.active_connections);
        println!("    Idle connections: {}", status.idle_connections);
        println!("    Waiting requests: {}", status.waiting_requests);
        println!("    Max connections: {}", status.max_connections);

        let utilization = if status.max_connections > 0 {
            (status.active_connections as f64 / status.max_connections as f64) * 100.0
        } else {
            0.0
        };
        println!("    Utilization: {:.1}%", utilization);
    } else {
        println!("  Connection pooling is not enabled");
    }
}

// Advanced example: Custom pool monitoring with metrics
#[allow(dead_code)]
async fn monitor_pool_with_metrics(client: &SmtpClient) -> Result<(), SmtpError> {
    println!("\n=== Pool Monitoring ===");

    // Spawn a background task to monitor pool health
    let client_clone = client.clone();
    let monitor_handle = tokio::spawn(async move {
        loop {
            if let Some(status) = client_clone.pool_status() {
                // Log metrics (in production, you'd send these to a monitoring system)
                println!("[MONITOR] Active: {}, Idle: {}, Waiting: {}",
                    status.active_connections,
                    status.idle_connections,
                    status.waiting_requests
                );

                // Alert if pool is exhausted
                if status.waiting_requests > 0 {
                    eprintln!("[ALERT] Pool exhausted! {} requests waiting",
                        status.waiting_requests);
                }

                // Alert if utilization is high
                let utilization = status.active_connections as f64 / status.max_connections as f64;
                if utilization > 0.8 {
                    println!("[WARNING] High pool utilization: {:.1}%", utilization * 100.0);
                }
            }

            sleep(Duration::from_secs(5)).await;
        }
    });

    // Let it run for a bit
    sleep(Duration::from_secs(30)).await;

    // Cancel monitoring
    monitor_handle.abort();

    Ok(())
}
